compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java

```diff
@@ -397,6 +406,170 @@ private void defineConstructs(BLangPackage pkgNode, SymbolEnv pkgEnv) {
                 .forEach(varSymbol -> varSymbol.tag = SymTag.ENDPOINT);
     }
 
+    private void defineErrorIntersection(SymbolEnv env) {
+        for (BLangNode typeDescriptor : this.errorIntersectionTypes) {
+            BLangTypeDefinition typeDefinition = (BLangTypeDefinition) typeDescriptor;
+
+            boolean isDistinctType = isDistinctFlagPresent(typeDefinition);
+            boolean isPublicType = typeDefinition.flagSet.contains(Flag.PUBLIC);
+            Set<BTypeIdSet.BTypeId> secondaryTypeIds = new HashSet<>();
+
+            BLangIntersectionTypeNode intersectionTypeNode = (BLangIntersectionTypeNode) typeDefinition.typeNode;
+            List<BLangType> constituentTypes = intersectionTypeNode.constituentTypeNodes;
+
+            BLangType bLangTypeOne = constituentTypes.get(0);
+            BType typeOne = symResolver.resolveTypeNode(bLangTypeOne, env);
+            if (typeOne.tag != TypeTags.ERROR && typeOne.tag != TypeTags.READONLY) {
+                dlog.error(bLangTypeOne.pos, INVALID_ERROR_INTERSECTION, bLangTypeOne);
+                continue;
+            }
+
+            BLangType bLangTypeTwo = constituentTypes.get(1);
+            BType typeTwo = symResolver.resolveTypeNode(bLangTypeTwo, env);
+            if (typeTwo.tag != TypeTags.ERROR && typeTwo.tag != TypeTags.READONLY) {
+                dlog.error(bLangTypeTwo.pos, INVALID_ERROR_INTERSECTION, bLangTypeTwo);
+                continue;
+            }
+
+            // Collecting typeIds.
+            if (typeOne.tag != TypeTags.READONLY) {
+                populateSecondaryTypeIdSet(secondaryTypeIds, (BErrorType) typeOne);
+            }
+            if (typeTwo.tag != TypeTags.READONLY) {
+                populateSecondaryTypeIdSet(secondaryTypeIds, (BErrorType) typeTwo);
+            }
+
+            BType potentialIntersectionType = types.getTypeIntersection(typeOne, typeTwo);
+
+            boolean isValidIntersection = true;
+            for (int i = 2; i < constituentTypes.size(); i++) {
+                BLangType bLangType = constituentTypes.get(i);
+                BType bType = symResolver.resolveTypeNode(bLangType, env);
+
+                if (bType.getKind() != TypeKind.ERROR && typeTwo.tag != TypeTags.READONLY) {
```

Comment:
Any specific reason to check kind instead of tag here?

Reply:
Fixed